src/Workspaces/Core/Portable/Serialization/PooledList.cs

```diff
@@ -15,7 +14,7 @@ namespace Microsoft.CodeAnalysis.Serialization
     /// </summary>
     internal static class Creator
     {
-        public static PooledObject<HashSet<Checksum>> CreateChecksumSet(ImmutableArray<Checksum> checksums)
+        public static PooledObject<HashSet<Checksum>> CreateChecksumSet(SegmentedList<Checksum> checksums)
```

Comment:
We don't want a mutable type here.

```suggestion
        public static PooledObject<HashSet<Checksum>> CreateChecksumSet(ImmutableSegmentedList<Checksum> checksums)
```

Reply:
I'd love for this to be an immutable type, but I don't see how to get one down to this level without allocating up in SynchronizeAssetsAsync. Maybe I'm missing a trick?

Reply:
@ToddGrun what about, as an alternative, a readonly struct wrapper around SegmentedList?

Reply:
I'm not clear on what that would gain. Yes, the wrapper would be immutable, but I assume it would give access to the mutable structure. I don't mind the current approach, especially since it shows promising allocation results, but am willing to consider other approaches.

Reply:
> a readonly struct wrapper around SegmentedList?

`ImmutableSegmentedList<T>` is exactly that.